CMake/FBThriftCppLibrary.cmake

```diff
@@ -0,0 +1,161 @@
+  endif()
+  set(thrift_include_options
+      "-I;$<JOIN:$<TARGET_PROPERTY:${LIB_NAME}.thrift_includes,INTERFACE_INCLUDE_DIRECTORIES>,;-I;>"
+  )
+
+  # Emit the rule to run the thrift compiler
+  add_custom_command(
+    OUTPUT ${generated_headers} ${generated_sources}
+    COMMAND_EXPAND_LISTS
+    COMMAND "${CMAKE_COMMAND}" -E make_directory "${output_dir}"
+    COMMAND
+      "${FBTHRIFT_COMPILER}"
+      # TODO: this flag does not exist in all fbthrift versions. --legacy-strict
+      --gen "mstch_cpp2:${GEN_ARG_STR}" "${thrift_include_options}" -I
+      "${FBTHRIFT_INCLUDE_DIR}" -o "${output_dir}"
+      "${CMAKE_CURRENT_SOURCE_DIR}/${THRIFT_FILE}"
+    WORKING_DIRECTORY "${CMAKE_BINARY_DIR}"
+    MAIN_DEPENDENCY "${THRIFT_FILE}"
+    DEPENDS ${ARG_DEPENDS} "${FBTHRIFT_COMPILER}")
+
+  # Now emit the library rule to compile the sources
+  if(BUILD_SHARED_LIBS)
+    set(LIB_TYPE SHARED)
+  else()
+    set(LIB_TYPE STATIC)
+  endif()
+
+  add_library("${LIB_NAME}" ${LIB_TYPE} ${generated_sources})
+
+  target_include_directories(
+    "${LIB_NAME}" PUBLIC "$<BUILD_INTERFACE:${CMAKE_BINARY_DIR}>"
+                         "$<INSTALL_INTERFACE:${ARG_INCLUDE_DIR}>")
+  target_link_libraries(
+    "${LIB_NAME}"
+    PUBLIC ${ARG_DEPENDS}
+           Folly::folly
+           # TODO: these symbols require other dependencies that need to be
+           # correctly handle by Velox's build system before they can be
+           # enabled. FBThrift::thriftcpp2 mvfst::mvfst_server_async_tran
+           # mvfst::mvfst_server
+  )
+
+  # Add ${generated_headers} to the PUBLIC_HEADER property for ${LIB_NAME}
+  #
+  # This allows callers to install it using "install(TARGETS ${LIB_NAME}
+  # PUBLIC_HEADER)" However, note that CMake's PUBLIC_HEADER behavior is rather
+  # inflexible, and does have any way to preserve header directory structure.
+  # Callers must be careful to use the correct PUBLIC_HEADER DESTINATION
+  # parameter when doing this, to put the files the correct directory
+  # themselves. We define a HEADER_INSTALL_DIR property with the include
+  # directory prefix, so typically callers should specify the PUBLIC_HEADER
+  # DESTINATION as "$<TARGET_PROPERTY:${LIB_NAME},HEADER_INSTALL_DIR>"
+  set_property(TARGET "${LIB_NAME}" PROPERTY PUBLIC_HEADER ${generated_headers})
+
+  # Define a dummy interface library to help propagate the thrift include
+  # directories between dependencies.
+  add_library("${LIB_NAME}.thrift_includes" INTERFACE)
+  target_include_directories(
+    "${LIB_NAME}.thrift_includes"
+    INTERFACE "$<BUILD_INTERFACE:${CMAKE_SOURCE_DIR}>"
+              "$<INSTALL_INTERFACE:${ARG_THRIFT_INCLUDE_DIR}>")
+  foreach(dep IN LISTS ARG_DEPENDS)
+    target_link_libraries("${LIB_NAME}.thrift_includes"
+                          INTERFACE "${dep}.thrift_includes")
+  endforeach()
+
+  set_target_properties(
+    "${LIB_NAME}"
+    PROPERTIES EXPORT_PROPERTIES "THRIFT_INSTALL_DIR"
```

Comment:
Not certain, but maybe @assignUser ought to know ; Is it right to export THRIFT_INSTALL_DIR etc ?

Reply:
same comment above.